### ISSUE TITLE
Offline MiniLLM defaults to a zero teacher signal

### ISSUE THREAD

**Issue by Amix29** [MakazhanAlpamys/Soup#692]
## Summary

Enabling MiniLLM with Soup's default settings selects the offline path with `minillm_teacher_mix_ratio=0.0`. In that path, the distillation target becomes the detached student distribution, so the MiniLLM term contains no teacher signal even though the teacher is loaded and forwarded.

Training does not necessarily look stuck because the ordinary cross-entropy term remains. The failure is therefore silent: the run can progress without performing the advertised teacher distillation.

## Root cause

Current `main` (`ef2e02b51a1b34546823edce55715d28234805cb`) defines:

```python
target = ratio * p_t + (1.0 - ratio) * p_s.detach()
```

while the schema defaults are:

```text
minillm_teacher_mix_ratio = 0.0
minillm_on_policy = false
```

The live distill wrapper passes that ratio into `MiniLLMConfig`. At zero, the offline loss is effectively:

```text
KL(p_student || stopgrad(p_student))
```

The teacher forward still happens before this calculation; its logits simply receive zero weight.

## Numerical reproduction

Using the actual `minillm_distill_term()` from the pinned checkout with fixed student logits and two different teachers:

| ratio | teacher A loss | teacher B loss | student gradient norm |
|---:|---:|---:|---:|
| 0.0 (default) | 0.0 | 0.0 | ~3.18e-17 |
| 1.0 (control) | 0.85786 | 0.73924 | ~0.129 |

At ratio zero, changing the teacher also leaves the student gradient unchanged. The normal cross-entropy term in `DistillTrainerWrapper` remains active; only the teacher-dependent MiniLLM signal disappears.

## Important scope distinction

A zero mix ratio is meaningful in the true on-policy path: it means sampling the rollout from the student, while the loss still compares student and teacher distributions. The degenerate behavior described here is specific to the default **offline distribution-blend** path.

The MiniLLM paper uses the mix ratio for the rollout sampling distribution while retaining a teacher-based reverse-KL objective: https://arxiv.org/html/2306.08543#S2.SS2.SSS3

## Recommended fix

Separate the sampling-mixture parameter from the offline distillation target. Viable short-term choices include:

1. Reject `minillm_enabled=true`, `minillm_on_policy=false`, and `teacher_mix_ratio=0` with a clear validation error; or
2. Give the offline path a non-zero teacher-dependent default; or
3. Define the offline reverse-KL target directly from the teacher and reserve `teacher_mix_ratio` for on-policy sampling.

Do not globally forbid zero because it remains valid for student-only sampling in the on-policy mode.

## Acceptance criteria

- [ ] The default enabled offline MiniLLM configuration produces a non-zero teacher-dependent term for non-identical teacher/student logits.
- [ ] Replacing the teacher logits changes the offline loss or gradient under the default enabled configuration.
- [ ] The on-policy path continues to allow ratio zero and still computes `KL(student || teacher)`.
- [ ] Configuration help clearly distinguishes sampling mix from offline objective behavior.
- [ ] A regression test asserts that the teacher cannot be silently ignored when MiniLLM is enabled offline.

## Duplicate check

#237 implemented the live helper; #257 added the optional true on-policy path; #263 optimized that rollout with KV caching. None tracks the enabled-by-default offline combination producing a zero teacher signal. Searches across open/closed issues and PRs found no matching report or open fix.

**Comment by MakazhanAlpamys** [comment]
**Confirmed. The default really is a no-op, and the teacher is loaded and forwarded anyway.**

Defaults:

```
minillm_teacher_mix_ratio = 0.0   minillm_on_policy = False   minillm_length_normalize = True
```

`minillm.py:220` is `target = ratio * p_t + (1.0 - ratio) * p_s.detach()`. At `ratio=0` the target **is** the detached student, so the reverse KL is identically zero and its gradient is exactly zero.

Kernel level:

```
ratio | teacher A loss | teacher B loss | |dLoss/dStudentLogits|_1
  0.0 |       0.000000 |       0.000000 | 7.381e-08   <-- SCHEMA DEFAULT
  0.3 |       0.049807 |       0.160424 | 1.979e-01
  1.0 |       0.625042 |       3.209177 | 9.118e-01
```

End to end through the real trainer, `minillm_enabled: true` and nothing else, against two teachers whose logits differ by roughly 22x:

```
config extras: (none - pure defaults)
MiniLLM distillation enabled (offline blend, teacher_mix=0.0, anchor=0.0)
teacher=A  mean|logit|=0.0540   total_loss=1.23754907   sum|grad|=0.00602429
teacher=B  mean|logit|=1.2005   total_loss=1.23754907   sum|grad|=0.00602429
```

**Bit-identical loss and bit-identical gradient from two completely different teachers.** Control at `ratio: 0.3` separates them (1.23770881 against 1.27182317).

## Your scope distinction checks out

`minillm_on_policy_rollout` accumulates `KL(student || teacher)` over full distributions and uses the ratio only for the per-token sampling coin-flip, so `ratio=0` is legitimate *there*. And the field's help text carries the on-policy semantics — *"Probability of sampling from the teacher distribution at rollout time. 0.0 = student-only"* — while the default path is the **offline objective blend**. One field, two meanings, and the default is only sensible under the one that is off by default. That is exactly the conflation you name.

## Why this one stings

The teacher is loaded, held in VRAM, and forwarded every step. The user pays the full memory and compute cost of distillation and receives a gradient of zero from it, while CE keeps the loss curve looking healthy. The console even prints `teacher_mix=0.0` without flagging it as degenerate.

There is already a footgun validator here — `_validate_minillm_compat` rejects setting `minillm_*` tunables *without* `minillm_enabled`. It has no rule for the inverse: enabled, offline, ratio 0. That asymmetry is the cheap fix, and it is in the file already as a pattern to copy.

## My ruling on which way to fix it

Two options and they are not equivalent. Changing the **default** to a non-zero ratio silently alters every existing `minillm_enabled` run's objective, which I do not want. **Refusing at config-parse time** — enabled, not on-policy, ratio 0 — costs nothing, names the exact field, and is consistent with how this project already handles degenerate combinations. Take that one unless you can argue the other.

Labelled `bug` + `help wanted`, unclaimed.

**Comment by jagadeepmamidi** [comment]
Hi @MakazhanAlpamys, I’d like to work on this. I’ll follow the suggested validation approach: reject the enabled offline MiniLLM configuration when minillm_teacher_mix_ratio is 0, while continuing to allow 0 for the on-policy path. I’ll also add regression tests covering both configurations and ensure the validation message names the relevant fields clearly. Could you please assign this to me?

**Comment by MakazhanAlpamys** [comment]
@jagadeepmamidi — **your plan is right, and I still cannot hand you this one yet. The reason is entirely my fault and you had no way to see it.**

## What actually happened

I offered **#692 and #693 to @SatvikMishra08 on 2026-09-06 at 04:56 UTC**, in the closing thread of #700, with a right of first refusal running to **2026-09-09 18:00 UTC**. You asked here at 12:08 UTC — seven hours later.

**But I never put the `claimed` label on either issue.** So for those seven hours, and for the twenty-four since, #692 has read as `bug` + `help wanted` with nobody on it. You did exactly the right thing against the state the repo was showing you. The state was wrong because I did not update it.

This is the fifth time in four days that I have held an issue in a comment and not in a label — #676, #699, #702, #700, and now this one. Both issues are labelled `claimed` as of now, which is the only form of a hold that shows up in `gh issue list`.

## Where that leaves this issue

- @SatvikMishra08 holds first refusal on #692 and #693 until **2026-09-09 18:00 UTC**.
- If either is still unclaimed after that, **you are first in line on #692**, on the strength of the plan you posted here. I will say so on this thread rather than making you ask again.

Your plan is also correct on the substance, including the part that is easy to get wrong: rejecting `minillm_teacher_mix_ratio: 0` only for the **offline** path, while `0` stays legal for `minillm_on_policy: true`. That distinction is the whole fix. Keep it.

## Something concrete now, if you want it

**#720 — distillation `compute_loss` ignores `num_items_in_batch` and can over-scale gradient accumulation.** Unclaimed, nobody has commented, and it is in `trainer/distill.py`, the same file family as #692, so nothing you have already read goes to waste. The issue carries a measured 4.0x ratio and names the mechanism: `model_accepts_loss_kwargs` is not set to `False`, so Transformers can skip its own accumulation compensation while Soup has not normalized by the global denominator.

Say the word and it is yours, with a `claimed` label this time — not just a comment.

You also still have **#689** (HF Hub `data.streaming` materialized eagerly) assigned. If two open items is one too many, take #720 later and I will hold it the same way.

**Comment by jagadeepmamidi** [comment]
Thanks for clarifying, @MakazhanAlpamys  no worries! I understand the hold on #692 and #693. I’d be happy to take #720 alongside #689; please mark it as claimed for me.

I’ll start by reproducing the gradient accumulation mismatch. Before implementing, could you confirm whether you’d prefer the minimal `model_accepts_loss_kwargs = False` fix first, or global token normalization as well? The acceptance criteria mention unequal microbatch lengths, so I’d like to make sure the scope matches what you expect.

**Comment by MakazhanAlpamys** [comment]
@jagadeepmamidi — answering the question you asked on 2026-09-07 so the thread is not left hanging, even though events have overtaken it.

The answer is **moot**: the minimal `model_accepts_loss_kwargs = False` fix is covered by **#742**, which is in review now. You stood down for this hold and took #720 and then #367 instead, which was the right call and cost you nothing — but it means the choice I owed you an answer on is no longer yours to make.

For completeness, had it still been live: **minimal fix first.** Global token normalization changes the loss every existing run produces, so it is a separate decision with its own reproduction burden, and bundling it would have made the small correct fix unreviewable.

Two things I owe you here, plainly:

1. **I held #692/#693 for @SatvikMishra08 on 2026-09-06 and turned you away for it.** I have just checked: they have no PRs in this repo, ever, and there has been no movement in seven days. I am pinging them now, and if there is no answer in a couple of days these reopen — and you have first refusal, because you stood down for the hold.
2. **You waited six days for a yes/no.** That is the second time this week I have found someone parked on a one-word answer from me. It is a process failure on my side, not a scheduling accident, and the fix is that I now answer direction questions before starting new review work rather than after.

**Comment by MakazhanAlpamys** [comment]
Fixed on `main` by @jagadeepmamidi in #977, merged as `3a37bf35`.

MiniLLM enabled with the default offline blend ignored the teacher while still loading it — at teacher mix 0 the offline reverse-KL target is the detached student, so the run paid for a teacher that contributed nothing. That combination is refused at config-parse time, and mix 0 stays legal on the on-policy path where it genuinely means student-only sampling.

The follow-up is the better half. `--minillm-on-policy` was applied by assigning `cfg.training.minillm_on_policy` **after** `load_config` had validated, and the schema declares no `validate_assignment`, so the new refusal fired before the flag was ever seen — telling a user who had just passed the flag to set the field the flag sets. Rather than moving the assignment, `load_config` gained a `training_overrides` parameter that merges into the raw `training:` mapping before `SoupConfig` is constructed, so a CLI flag now passes through the same cross-validators as YAML.

```
soup train --config soup.yaml --minillm-on-policy   # proceeds
soup train --config soup.yaml                       # still refuses
```

Both directions verified by running them. The regression test invokes the CLI rather than grepping `--help`, which matters because a `--help` assertion would have passed against the broken code.

Credit to @jagadeepmamidi, who turned the review round around in under an hour.
